src/contracts/atlas/ExecutionEnvironment.sol

```diff
@@ -121,179 +110,117 @@ contract ExecutionEnvironment is Base {
     }
 
-    /// @notice The solverMetaTryCatch function is called by Atlas to execute the SolverOperation, as well as any
-    /// preSolver or postSolver hooks that the DAppControl contract may require.
-    /// @dev This contract is called by the Atlas contract, delegatecalls the preSolver and postSolver hooks if
-    /// required, and executes the SolverOperation by calling the `solverOp.solver` address.
+    /// @notice The solverPreTryCatch function is called by Atlas to execute the preSolverCall part of each
+    /// SolverOperation. A SolverTracker struct is also returned, containing bid info needed to handle the difference in
+    /// logic between inverted and non-inverted bids.
     /// @param bidAmount The Solver's bid amount.
-    /// @param gasLimit The gas limit for the SolverOperation.
     /// @param solverOp The SolverOperation struct.
     /// @param returnData Data returned from the previous call phase.
-    function solverMetaTryCatch(
+    /// @return solverTracker Bid tracking information for the current solver.
+    function solverPreTryCatch(
         uint256 bidAmount,
-        uint256 gasLimit,
         SolverOperation calldata solverOp,
         bytes calldata returnData
     )
         external
         payable
-        onlyAtlasEnvironment(ExecutionPhase.SolverOperations, _ENVIRONMENT_DEPTH)
+        onlyAtlasEnvironment(ExecutionPhase.PreSolver, _ENVIRONMENT_DEPTH)
+        returns (SolverTracker memory solverTracker)
     {
-        if (address(this).balance != solverOp.value) revert AtlasErrors.SolverMetaTryCatchIncorrectValue();
-
-        uint32 config = _config();
-        address control = _control();
-
-        // Track token balance to measure if the bid amount is paid.
-        bool etherIsBidToken;
-        uint256 startBalance;
-
-        if (solverOp.bidToken == address(0)) {
-            startBalance = 0; // address(this).balance - solverOp.value;
-            etherIsBidToken = true;
-            // ERC20 balance
-        } else {
-            startBalance = ERC20(solverOp.bidToken).balanceOf(address(this));
-        }
-
-        ////////////////////////////
-        // SOLVER SAFETY CHECKS //
-        ////////////////////////////
-
         // Verify that the DAppControl contract matches the solver's expectations
-        if (solverOp.control != control) {
+        if (solverOp.control != _control()) {
             revert AtlasErrors.AlteredControl();
         }
 
-        bool success;
+        solverTracker.bidAmount = bidAmount;
+        solverTracker.etherIsBidToken = solverOp.bidToken == address(0);
+
+        // bidValue is inverted; Lower bids are better; solver must withdraw <= bidAmount
+        if (_config().invertsBidValue()) {
+            solverTracker.invertsBidValue = true;
+            // if invertsBidValue, record ceiling now
+            // inventory to send to solver must have been transferred in by userOp or preOp call
```

Comment:
But how do we know that the EE had a balance of 0 for the bid token before the preOp?

Reply:
I think this accommodates zero or non-zero starting balances in the EE of either ETH or ERC20 assets. It does this by only looking at the change in the bid asset between pre and post solver hook calls - i.e. the change in balance due to the actual solver call that happens in between. This change is the solver's bid.